private/buf/cmd/buf/command/push/push.go

```diff
@@ -347,6 +386,120 @@ func validateLabelFlagValues(flags *flags) error {
 	return nil
 }
 
+// We do not allow users to set --source-control-url, --create-default-label, and --label
+// flags if the --git-metadata flag is set.
+func validateGitMetadataFlags(flags *flags) error {
+	if flags.GitMetadata {
+		var usedFlags []string
+		if flags.SourceControlURL != "" {
+			usedFlags = append(usedFlags, sourceControlURLFlagName)
+		}
+		if flags.CreateDefaultLabel != "" {
+			usedFlags = append(usedFlags, createDefaultLabelFlagName)
+		}
+		if len(flags.Tags) > 0 {
+			usedFlags = append(usedFlags, tagFlagName)
+		}
+		if flags.Branch != "" {
+			usedFlags = append(usedFlags, branchFlagName)
+		}
+		if flags.Draft != "" {
+			usedFlags = append(usedFlags, draftFlagName)
+		}
+		if len(usedFlags) > 0 {
+			usedFlagsErrStr := strings.Join(
+				slicesext.Map(
+					usedFlags,
+					func(flag string) string { return fmt.Sprintf("--%s", flag) },
+				),
+				", ",
+			)
+			return appcmd.NewInvalidArgumentErrorf(
+				"The following flag(s) cannot be used in combination with --%s: %s.",
+				gitMetadataFlagName,
+				usedFlagsErrStr,
+			)
+		}
+	}
+	return nil
+}
+
+func getGitMetadataUploadOptions(
+	ctx context.Context,
+	container appext.Container,
+	flags *flags,
+) ([]bufmodule.UploadOption, error) {
+	input, err := bufcli.GetInputValue(container, flags.InputHashtag, ".")
+	if err != nil {
+		return nil, err
+	}
+	// validate that input is a local directory
+	fileInfo, err := os.Stat(input)
+	if err != nil || !fileInfo.IsDir() {
+		return nil, appcmd.NewInvalidArgumentErrorf("input, %s, must be a local Git repository checkout: %w", input, err)
+	}
+	runner := command.NewRunner()
+	uncommittedFiles, err := git.CheckForUncommittedGitChanges(ctx, runner, input)
+	if err != nil {
+		return nil, err
+	}
+	if len(uncommittedFiles) > 0 {
+		return nil, fmt.Errorf("uncommitted changes found in the following files: %v", uncommittedFiles)
+	}
+	originRemote, err := git.GetRemote(ctx, runner, container, input, gitOriginRemote)
+	if err != nil {
+		if errors.Is(err, git.ErrRemoteNotFound) {
+			return nil, appcmd.NewInvalidArgumentErrorf("remote %s must be present on Git checkout: %w", gitOriginRemote, err)
+		}
+		return nil, err
+	}
+	currentGitCommit, err := git.GetCurrentHEADGitCommit(ctx, runner, input)
+	if err != nil {
+		return nil, err
+	}
+	var gitMetadataUploadOptions []bufmodule.UploadOption
+	gitLabelsUploadOption, err := getGitMetadataLabelsUploadOptions(ctx, runner, container, input, currentGitCommit)
+	if err != nil {
+		return nil, err
+	}
+	if gitLabelsUploadOption != nil {
+		gitMetadataUploadOptions = append(gitMetadataUploadOptions, gitLabelsUploadOption)
+	}
+	sourceControlURL := originRemote.SourceControlURL(currentGitCommit)
+	if sourceControlURL == "" {
+		return nil, appcmd.NewInvalidArgumentError("unable to determine source control URL for this repository; only GitHub/GitLab/BitBucket are supported")
```

Comment:
Why does it matter? This is a severe limitation, and excludes any github enterprise customer. We should be able to parse a source control url for any http/https/ssh endpoint

Reply:
There are two potential questions here so will answer both.

**Why is it an error if we can't infer a source control url?**
If the user provided `--git-metadata` then they are expecting us to automatically set some values. If we can't set the values they are expecting, we should error. This means during setup, they will catch the problem early, and then later if, due to a code host migration, it breaks, they will know instantly.

**Why are we limited to GitHub/Bitbucket/GitLab?**
The specification for Remote.SourceControlURL() (which is in the interface documentation in this PR) is designed to work for enterprise customers as long as the name of their code host is in the remote name. If it isn't, we don't know what code host we are dealing with and therefore we don't know how to correctly construct the user facing url because the routes are different depending on the code host.

On GitHub and GitLab, routes look like `https://<host><repo-path>/commit/<commit-sha>`
and on BitBucket, routes look like `https://<host><repo-path>/commits/<commit-sha>`
(note `/commit/` vs `/commits`).

Reply:
The code quite literally checks for if the hostname includes `github, gitlab, bitbucket`, and just with a `strings.Contains` on the hostname. This both requires hostnames to include github, for example (this isn't a hard requirement), and it would also accept things like `foogithubbar`, which doesn't feel like what this filter wants to do in the first place.

This limitation isn't acceptable for our enterprise customers. Additionally, my impression was that "source control URL" wasn't mean to link to a specific point-in-time on the repository, it was just meant to link to the repository itself.

We can discuss deferring this until after release, but I consider this a bug - limiting acceptable hostnames to `strings.Contains` on one of `github, gitlab, bitbucket` isn't acceptable